runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java

```diff
@@ -0,0 +1,226 @@
+     * WARNING: The ID returned by this function will NOT be the comparable with the ID
+     *          returned by the CorfuQueueRecord from entryList() method because if this
+     *          method executes within a transaction, the comparable ordering is not
+     *          known until the transaction commits.
+     *          The ID returned here is only really useful for remove() operations.
+     */
+    public UUID enqueue(E e) {
+        final Long id = guidGenerator.nextLong();
+        corfuTable.put(id, e);
+        return new UUID(0, id);
+    }
+
+    /**
+     * CorfuQueueRecord encapsulates each entry enqueued into CorfuQueue with its unique ID.
+     * It is a read-only type returned by the entryList() method.
+     * The ID returned here can be used for both point get()s as well as remove() operations
+     * on this Queue.
+     *
+     * @param <E>
+     */
+    public static class CorfuQueueRecord<E> {
+        /**
+         * This ID represents the entry and its order in the Queue.
+         * This implies that it is unique and comparable with other IDs
+         * returned from CorfuQueue methods with respect to its enqueue order.
+         * However it cannot be compared with ID returned from the enqueue method for ordering
+         * because if this method is wrapped in a transaction, the order is established only later.
+         */
+        @Getter
+        final UUID ID;
+
+        @Getter
+        final E entry;
+
+        public String toString() {
+            return String.format(ID.toString()+"=>"+entry.toString());
+        }
+
+        CorfuQueueRecord(UUID id, E entry) {
+            this.ID = id;
+            this.entry = entry;
+        }
+    }
+
+    final private static int MAX_BITS_FOR_INDEX = 24;
+    /**
+     * Returns a List of CorfuQueueRecords sorted by the order in which the enqueue materialized.
+     * This is the primary method of consumption of entries enqueued into CorfuQueue.
+     *
+     * To re-constitute the commit order in the presence of transactions the CorfuRecords returned
+     * here have their UUID fields composed of two parts:
+     *     +----------------------------------------------------------------------------+
+     *     | Stream snapshot |Index in snapshot|     ID of the entry in the map        |
+     *     +----------------------------------------------------------------------------+
+     *     <----5 bytes------><----3 bytes-----><------- 8 bytes ----------------------->
+     *
+     * <p>This function currently does not return a view like the java.util implementation,
+     * and changes to the entryList will *not* be reflected in the map. </p>
+     *
+     * @param maxEntries - Limit the number of entries returned from start of the queue
+     * @return List of Entries sorted by their enqueue order
+     */
+    public List<CorfuQueueRecord<E>> entryList(int maxEntries) {
+        if (maxEntries < 0) {
+            maxEntries = Integer.MAX_VALUE;
```

Comment:
please add logs - suspicious value. Why it less than zero?

Reply:
done. better to throw illegal exception I think

Reply:
My advice is  instead of throwing an Exception use https://github.com/CorfuDB/CorfuDB/blob/master/common/src/main/java/org/corfudb/common/result/Result.java